simple/reprovide.go

```diff
@@ -96,28 +124,41 @@ func (rp *Reprovider) Reprovide() error {
 	if err != nil {
 		return fmt.Errorf("failed to get key chan: %s", err)
 	}
-	for c := range keychan {
-		// hash security
-		if err := verifcid.ValidateCid(c); err != nil {
-			logR.Errorf("insecure hash in reprovider, %s (%s)", c, err)
-			continue
-		}
-		op := func() error {
-			err := rp.rsys.Provide(rp.ctx, c, true)
-			if err != nil {
-				logR.Debugf("Failed to provide key: %s", err)
-			}
-			return err
-		}
 
-		// TODO: this backoff library does not respect our context, we should
-		// eventually work contexts into it. low priority.
-		err := backoff.Retry(op, backoff.NewExponentialBackOff())
-		if err != nil {
-			logR.Debugf("Providing failed after number of retries: %s", err)
-			return err
-		}
+	var wg sync.WaitGroup
+
+	for workers := 0; workers < rp.workerLimit; workers++ {
```

Comment:
I'm not sure I understand the idea of this PR. It looks like it's going to publish the same keys 8 times.

Reply:
It seems to work as intended: https://play.golang.org/p/xXbMVNzSGtp

Reply:
To be more explicit: all the worker share the same source channel. When they pick a value, it's not available for the others anymore, effectively fanning out the work on the workers.

Reply:
Right I see, sorry I had missed that it's a channel

Reply:
Could you please add a test for this new behaviour?

Reply:
Well, with the PR the default is now to republish concurrently so the tests already cover the new behavior. Admittedly, publishing only 4 blocks is not exactly realistic though. I'll update them.

Reply:
Done. I also found and fixed a bug in the tests, they were not actually testing much (see where `var providers []peer.AddrInfo` is declared).

Reply:
Thanks for fixing that 👍